src/components/structures/RoomStatusBar.tsx

```diff
@@ -265,6 +291,51 @@ export default class RoomStatusBar extends React.PureComponent<IProps, IState> {
         );
     }
 
+    private getUpdatedAcknowledgedHistoryVisibility(): boolean {
+        let acknowledgedHistoryVisibility = SettingsStore.getValue(
+            "acknowledgedHistoryVisibility",
+            this.props.room.roomId,
+        );
+        if (this.props.room.getHistoryVisibility() === HistoryVisibility.Joined) {
+            acknowledgedHistoryVisibility = false;
+            // Clear the dismissed flag to ensure if a room is changed public -> private -> public,
+            // we show the banner again when it is set back to public.
+            void SettingsStore.setValue(
+                "acknowledgedHistoryVisibility",
+                this.props.room.roomId,
+                SettingLevel.ROOM_ACCOUNT,
+                false,
+            );
+        }
+        return acknowledgedHistoryVisibility;
+    }
+
+    private shouldShowHistoryVisibilityContent(): boolean {
+        return (
+            this.state.roomHasEncryptionStateEvent &&
+            this.state.roomHistoryVisibility !== HistoryVisibility.Joined &&
+            !this.state.acknowledgedHistoryVisibility
+        );
+    }
+
+    private getHistoryVisibilityContent(): JSX.Element {
+        return (
+            <RoomStatusBarHistoryVisible
+                onClose={async () => {
+                    await SettingsStore.setValue(
+                        "acknowledgedHistoryVisibility",
+                        this.props.room.roomId,
+                        SettingLevel.ROOM_ACCOUNT,
+                        true,
+                    );
+                    this.setState({
+                        acknowledgedHistoryVisibility: true,
+                    });
```

Comment:
What is happenning if the promise fails?

Reply:
Hmm, a good question. We could set the component's state regardless of whether the promise fails? This would mean that if the user can't reach the server and switches rooms, the banner will reappear when it shouldn't.